### PR TITLE
Support building multi-ASIC components

### PR DESCRIPTION
**Pull request by theasianpianist**
**- What I did**
* Move single instance services into their own folder
* Generate Systemd templates for any multi-instance service files in slave.mk
* detect single or multi-instance platform in systemd-sonic-generator based on onie_platform in /host/machine.conf
* add Systemd service to create network namespaces
* update container hostname after creation instead of during creation (docker_image_ctl)
* run Docker containers in a network namespace if specified
* add a service to create a simulated multi-ASIC topology on the virtual switch platform

**- How to verify it**
Build should succeed - running the image on a single-ASIC platform should perform normally. 
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: dockers/docker-namespace/Dockerfile.j2 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+FROM debian:stretch
```
Comment (lguohan) [inline dockers/docker-namespace/Dockerfile.j2]: why do we need this docker?

File: platform/vs/sonic.xml [hunk 2]
```diff
@@ -22,7 +22,7 @@
     <emulator>/usr/bin/qemu-system-x86_64</emulator>
     <disk type='file' device='disk'>
       <driver name='qemu' type='qcow2' cache='writeback'/>
-      <source file='/data/sonic/sonic-buildimage/target/sonic-vs.img'/>
+      <source file='/home/t-lale/sonic-buildimage/target/sonic-vs.img'/>
```
Comment (lguohan) [inline platform/vs/sonic.xml]: do not modify this line.

Comment (lguohan) [inline platform/vs/sonic.xml]: remove this file change, it is not neccessary.

File: slave.mk [hunk 3]
```diff
@@ -535,7 +535,19 @@ $(addprefix $(TARGET_PATH)/, $(DOCKER_IMAGES)) : $(TARGET_PATH)/%.gz : .platform
...
+		--build-arg http_proxy=$(HTTP_PROXY) \
+		--build-arg https_proxf=$(HTTPS_PROXY) \
+		--build-arg user=$(USER) \
+		--build-arg uid=$(UID) \
+		--build-arg guid=$(GUID) \
+		--build-arg docker_container_name=$($*.gz_CONTAINER_NAME) \
+		--build-arg frr_user_uid=$(FRR_USER_UID) \
+		--build-arg frr_user_gid=$(FRR_USER_GID) \
+		--label Tag=$(SONIC_GET_VERSION) \
+		-t $* $($*.gz_PATH) $(LOG)
```
Comment (lguohan) [inline slave.mk]: why this change?

File: slave.mk [hunk 4]
```diff
@@ -632,22 +644,47 @@ $(addprefix $(TARGET_PATH)/, $(SONIC_INSTALLERS)) : $(TARGET_PATH)/% : \
 	export platform_common_py2_wheel_path="$(addprefix $(PYTHON_WHEELS_PATH)/,$(SONIC_PLATFORM_COMMON_PY2))"
 	export redis_dump_load_py2_wheel_path="$(addprefix $(PYTHON_WHEELS_PATH)/,$(REDIS_DUMP_LOAD_PY2))"
 	export install_debug_image="$(INSTALL_DEBUG_TOOLS)"
+
```
Comment (lguohan) [inline slave.mk]: do not change.

File: dockers/docker-namespace/Dockerfile.j2 (hunk 1 above)
Comment (lguohan) [reply dockers/docker-namespace/Dockerfile.j2]: we can use the database as the namespace container, all other containers will use the same namespace as the database container. we do not need a separate namespace container.

File: src/systemd-sonic-generator/systemd-sonic-generator.c [hunk 5]
```diff
@@ -12,10 +12,12 @@
 #define MAX_NUM_TARGETS 5
 #define MAX_NUM_INSTALL_LINES 5
 #define MAX_NUM_UNITS 128
+#define NUM_ASICS 6
```
Comment (lguohan) [inline src/systemd-sonic-generator/systemd-sonic-generator.c]: why hard code here?

File: rules/docker-namespace.mk [hunk 6]
```diff
@@ -0,0 +1,13 @@
+DOCKER_NAMESPACE = docker-namespace.gz
```
Comment (lguohan) [inline rules/docker-namespace.mk]: we should remove this file

File: files/image_config/topology/topology.sh [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    # eth48 - eth63: namespace5
+    for NS in `seq 2 5`; do
+        for NUM in `seq 1 16`; do
+            ORIG="eth$((16 * $(($NS - 2)) + $NUM - 1))"
+            TEMP="ethTemp999"
+            NEW="eth$(($NUM + 16))"
+            ip link set dev $ORIG down
+            ip link set dev $ORIG name $TEMP # rename to prevent conflicts before renaming in new namespace
+            ip link set dev $TEMP netns namespace$NS
+            sudo ip netns exec namespace$NS ip link set $TEMP name $NEW # rename to final interface name
```
Comment (lguohan) [inline files/image_config/topology/topology.sh]: rename namespace to asic$dev?

File: files/image_config/topology/topology.sh [hunk 8]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+start () {
+    # Move external links into assigned frontend namespaces
+    # eth0  - eth15: namespace2
+    # eth16 - eth31: namespace3
+    # eth32 - eth47: namespace4
+    # eth48 - eth63: namespace5
+    for NS in `seq 2 5`; do
```
Comment (lguohan) [inline files/image_config/topology/topology.sh]: NS-> ASIC

File: src/systemd-sonic-generator/systemd-sonic-generator.c (hunk 5 above)
Comment (theasianpianist) [reply src/systemd-sonic-generator/systemd-sonic-generator.c]: I'm not really sure how else to get this information to the generator - later on it can probably read from a file somewhere.

File: dockers/docker-namespace/Dockerfile.j2 (hunk 1 above)
Comment (theasianpianist) [reply dockers/docker-namespace/Dockerfile.j2]: I have this since I start the namespaces, then the topology service is started, then the other containers start. I wanted to have the topology set up before any SONiC containers were started.

File: slave.mk (hunk 3 above)
Comment (theasianpianist) [reply slave.mk]: This is needed if we want to keep the namespace container, since the docker build fails when it tries to squash the container

Comment (lguohan) [reply slave.mk]: so, it is tight to namespace container, let me resolve this in the conversation above.

File: dockers/docker-namespace/Dockerfile.j2 (hunk 1 above)
Comment (lguohan) [reply dockers/docker-namespace/Dockerfile.j2]: topology service is only required for kvm image. it is not necessary to have topology build before all containers start. We can build topology after database docker started. in fact, the bgp session will not be able to setup until host interfaces are created.

the only hard requirement I see is that the peer links has to be start before the syncd docker starts. since syncd will link the peer device  with the host interface.
